lib/src/Pos/Block/Logic/Header.hs

```diff
@@ -353,34 +353,67 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
--- CSL-396 don't load all the blocks into memory at once
 -- | Given @from@ and @to@ headers where @from@ is older (not strict)
 -- than @to@, and valid chain in between can be found, headers in
 -- range @[from..to]@ will be found.
-getHeadersFromToIncl
+getHeadersRange
     :: forall m. (HasConfiguration, MonadDBRead m)
-    => HeaderHash -> HeaderHash -> m (Maybe (OldestFirst NE HeaderHash))
-getHeadersFromToIncl older newer = runMaybeT . fmap OldestFirst $ do
+    => HeaderHash -> HeaderHash -> m (Either Text (OldestFirst NE HeaderHash))
+getHeadersRange older newer | older == newer = runExceptT $ do
+    unlessM (isJust <$> DB.getHeader newer) $
+        throwError "getHeadersRange: can't find newer-older header"
+    pure $ OldestFirst $ one newer
+getHeadersRange older newer = runExceptT $ do
     -- oldest and newest blocks do exist
-    start <- MaybeT $ DB.getHeader newer
-    end   <- MaybeT $ DB.getHeader older
-    guard $ getEpochOrSlot start >= getEpochOrSlot end
-    let lowerBound = getEpochOrSlot end
-    if newer == older
-    then pure $ one newer
-    else loadHeadersDo lowerBound (one newer) $ start ^. prevBlockL
+    newerHd <- fromMaybeM "can't retrieve newer header" $ DB.getHeader newer
+    olderHd <- fromMaybeM "can't retrieve older header" $ DB.getHeader older
+    let olderD = olderHd ^. difficultyL
+    let newerD = newerHd ^. difficultyL
+
+    -- Proving newerD >= olderD
+    when (newerD == olderD) $
+        throwError $ "getHeadersRange: newer and older headers have " <>
+                     "the same difficulty, but are not equal"
+    when (newerD < olderD) $
+        throwError "getHeadersRange: newer header is less dificult that older one"
+
+    -- How many epochs does this range cross.
+    let genDiff :: Int
+        genDiff = fromIntegral $ newerHd ^. epochIndexL - olderHd ^. epochIndexL
+    -- Number of blocks is difficulty difference + number of genesis blocks.
+    let depthDiff :: Int
+        depthDiff = genDiff + fromIntegral (newerD - olderD)
+
+    -- We load these depthDiff blocks.
+    let cond _curHash depth = depth < depthDiff
```

Comment:
Omg, what's the purpose of doing all this stuff with difficulty (which looks rather complicated), when you can simply check whether `_curHash == newer`?

Reply:
Because i don't want to traverse more than `depthDiff` blocks ever. Replacing the check with the one you suggest would open an DOS attack vector.

Reply:
I don't get it. Currently this function can load all blocks, there is no upper bound other than that. If you pass `const (const True)` as `cond`, there will be the same bound.

Can you elaborate on DOS attack vector which your solution makes impossible?

Reply:
Oh, indeed, sorry -- I assumed that nobody can query more than `recoveryHeadersMessage` blocks at once. I guess i should put this limitation there.

Attack is basically asking to send all the chain, from the very first block, to the last one. When i put limitation on the `depthDiff` it will be better.

Indeed, `cond` can be expressed both in terms of depth (how it is done now) and it rems of `curHash == newer`. There's no difference.

Reply:
I still think that it would be simpler, if `cond` was checking `curHash == newer && maybe True (depth <=) depthLimitM` rather than manipulating difficulties. But probably it's subjective, so I won't continue this discussion.